Replace binary Analysis format with a text-based one.

benjyw opened the pull request:
Reads/writes are a little faster with the text format,
and it's far more useful. E.g., it allows external manipulation
and inspection of the analysis.

We don't gzip the output. It does greatly shrink the files,
however it makes reads and writes 1.5x-2x slower, and we're
optimizing for speed over compactness.

Commit:
Replace binary Analysis format with a text-based one.
Reads/writes are a little faster with the text format,
and it's far more useful. E.g., it allows external manipulation
and inspection of the analysis.

We don't gzip the output. It does greatly shrink the files,
however it makes reads and writes 1.5x-2x slower, and we're
optimizing for speed over compactness.

gkossakowski commented:
Hi Benjy!

Thanks for submitting the patch! I'm reviewing it right now.

W dniu czwartek, 14 listopada 2013 użytkownik Benjy Weinberger napisał:

> Reads/writes are a little faster with the text format,
> and it's far more useful. E.g., it allows external manipulation
> and inspection of the analysis.
> 
> We don't gzip the output. It does greatly shrink the files,
> however it makes reads and writes 1.5x-2x slower, and we're
> 
> ## optimizing for speed over compactness.
> 
> You can merge this Pull Request by running
> 
>   git pull https://github.com/benjyw/sbt text_analysis_file
> 
> Or view, comment on, or merge it at:
> 
>   https://github.com/sbt/sbt/pull/985
> Commit Summary
> - Replace binary Analysis format with a text-based one.
> 
> File Changes
> - _M_
>   compile/integration/src/main/scala/sbt/compiler/AggressiveCompile.scalahttps://github.com/sbt/sbt/pull/985/files#diff-0(2)
> - _M_
>   compile/integration/src/main/scala/sbt/compiler/IncrementalCompiler.scalahttps://github.com/sbt/sbt/pull/985/files#diff-1(11)
> - _M_ compile/persist/src/main/scala/sbt/inc/AnalysisFormats.scalahttps://github.com/sbt/sbt/pull/985/files#diff-2(2)
> - _M_ compile/persist/src/main/scala/sbt/inc/FileBasedStore.scalahttps://github.com/sbt/sbt/pull/985/files#diff-3(16)
> - _D_ compile/persist/src/main/scala/sbt/inc/InternPool.scalahttps://github.com/sbt/sbt/pull/985/files#diff-4(82)
> - _D_
>   compile/persist/src/main/scala/sbt/inc/InternedAnalysisFormats.scalahttps://github.com/sbt/sbt/pull/985/files#diff-5(98)
> - _A_ compile/persist/src/main/scala/sbt/inc/TextAnalysisFormat.scalahttps://github.com/sbt/sbt/pull/985/files#diff-6(312)
> 
> Patch Links:
> - https://github.com/sbt/sbt/pull/985.patch
> - https://github.com/sbt/sbt/pull/985.diff

## 

Grzegorz Kossakowski
Scalac hacker at Typesafe http://www.typesafe.com/
twitter: @gkossakowski http://twitter.com/gkossakowski
github: @gkossakowski http://github.com/gkossakowski

gkossakowski commented on compile/integration/src/main/scala/sbt/compiler/IncrementalCompiler.scala:
```diff
@@ -40,10 +40,5 @@ object IC extends IncrementalCompiler[Analysis, AnalyzingCompiler]
-		import formats._
-		sbt.IO.gzipFileIn(file)( in => sbinary.Operations.read[(Analysis, CompileSetup)](in) )
-	}
+		sbt.IO.gzipFileIn(file)( in => TextAnalysisFormat.read(new BufferedReader(new InputStreamReader(in))) )
```
You mention in commit message that gzip is not used anymore. Why is it used here?

benjyw replied:
Ooops, missed that. Just pushed a commit to fix it.

On Thu, Nov 14, 2013 at 10:24 AM, Grzegorz Kossakowski <
notifications@github.com> wrote:

> In
> compile/integration/src/main/scala/sbt/compiler/IncrementalCompiler.scala:
> 
> > @@ -40,10 +40,5 @@ object IC extends IncrementalCompiler[Analysis, AnalyzingCompiler]
> >         try { readCacheUncaught(file)._1 } catch { case _: Exception => Analysis.Empty }
> > 
> > ```
> > def readCacheUncaught(file: File): (Analysis, CompileSetup) =
> > ```
> > -   {
> > -       import sbinary.DefaultProtocol.{immutableMapFormat, immutableSetFormat, StringFormat, tuple2Format}
> > -       val formats = new sbt.inc.InternedAnalysisFormats()
> > -       import formats._
> > -       sbt.IO.gzipFileIn(file)( in => sbinary.Operations.read[(Analysis, CompileSetup)](in) )
> > -   }
> > -       sbt.IO.gzipFileIn(file)( in => TextAnalysisFormat.read(new BufferedReader(new InputStreamReader(in))) )
> 
> You mention in commit message that gzip is not used anymore. Why is it
> used here?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/sbt/sbt/pull/985/files#r7668127
> .

Commit:
Remove a stray gzip use.

gkossakowski commented on compile/persist/src/main/scala/sbt/inc/TextAnalysisFormat.scala:
```diff
@@ -0,0 +1,312 @@
+
+	def close(key: String) {
+		if (printTimings) {
+			println("[%s] %dms".format(key, timers.getOrElse(key, 0L) / 1000000))
```
I was thinking whether we would like to use proper logging API here but I'm guessing that just for printing timings this should be ok.

benjyw replied:
Well, this only prints when that sysprop is set, and I want to get rid of
this anyway, once we've convinced ourselves that the code is reasonably
optimal. I had used these timers to squeeze performance gains out anywhere
I could, but the main remaining issue is the sbinary serialization of the
API objects...

On Thu, Nov 14, 2013 at 3:52 PM, Grzegorz Kossakowski <
notifications@github.com> wrote:

> In compile/persist/src/main/scala/sbt/inc/TextAnalysisFormat.scala:
> 
> > -       val start = System.nanoTime()
> > -       val ret = f
> > -       val elapsed = System.nanoTime() - start
> > -       timers.update(key, timers.getOrElseUpdate(key, 0) + elapsed)
> > -       ret
> > -   }
> >   +
> > -   def time[T](key: String)(f: => T) = {
> > -       val ret = aggregate(key)(f)
> > -       close(key)
> > -       ret
> > -   }
> >   +
> > -   def close(key: String) {
> > -       if (printTimings) {
> > -           println("[%s] %dms".format(key, timers.getOrElse(key, 0L) / 1000000))
> 
> I was thinking whether we would like to use proper logging API here but
> I'm guessing that just for printing timings this should be ok.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/sbt/sbt/pull/985/files#r7677978
> .

gkossakowski commented on compile/persist/src/main/scala/sbt/inc/TextAnalysisFormat.scala:
```diff
@@ -0,0 +1,312 @@
+
+	object RelationsF {
+		object Headers {
+			val srcProd = "products"
```
Do we want this to be public?

gkossakowski commented on compile/persist/src/main/scala/sbt/inc/TextAnalysisFormat.scala:
```diff
@@ -0,0 +1,312 @@
+		(Analysis.Empty.copy(stamps, apis, relations, infos, compilations), setup)
+	}
+
+	object RelationsF {
```
@benjyw, do you intentionally made this object and all inner objects public?

benjyw replied:
No, they could all be private.

On Thu, Nov 14, 2013 at 3:53 PM, Grzegorz Kossakowski <
notifications@github.com> wrote:

> In compile/persist/src/main/scala/sbt/inc/TextAnalysisFormat.scala:
> 
> > -       FormatTimer.time("write setup") { CompileSetupF.write(out, setup) }
> > -       out.flush()
> > -   }
> >   +
> > -   def read(in: BufferedReader): (Analysis, CompileSetup) = {
> > -       val relations = FormatTimer.time("read relations") { RelationsF.read(in) }
> > -       val stamps = FormatTimer.time("read stamps") { StampsF.read(in) }
> > -       val apis = FormatTimer.time("read apis") { APIsF.read(in) }
> > -       val infos = FormatTimer.time("read sourceinfos") { SourceInfosF.read(in) }
> > -       val compilations = FormatTimer.time("read compilations") { CompilationsF.read(in) }
> > -       val setup = FormatTimer.time("read setup") { CompileSetupF.read(in) }
> >   +
> > -       (Analysis.Empty.copy(stamps, apis, relations, infos, compilations), setup)
> > -   }
> >   +
> > -   object RelationsF {
> 
> @benjyw https://github.com/benjyw, do you intentionally made this
> object and all inner objects public?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/sbt/sbt/pull/985/files#r7678026
> .

gkossakowski commented:
Overall, looks very good to me. Once questions related to access of newly introduced apis are addressed I think we can merge it.

I tested it on a few projects (scalacheck, scala compiler, scalatest) and it worked well everywhere.

Great work @benjyw! I wanted text-based format for Analysis for a long time!

Commit:
Make text format impl details private.

I used private[this] out of habit, even though the scope is
an object, so just private would have the same effect.

benjyw commented:
Pushed a commit that makes all the inner objects private[this].

On Thu, Nov 14, 2013 at 3:57 PM, Grzegorz Kossakowski <
notifications@github.com> wrote:

> Overall, looks very good to me. Once questions related to access of newly
> introduced apis are addressed I think we can merge it.
> 
> I tested it on a few projects (scalacheck, scala compiler, scalatest) and
> it worked well everywhere.
> 
> Great work @benjyw https://github.com/benjyw! I wanted text-based
> format for Analysis for a long time!
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/sbt/sbt/pull/985#issuecomment-28535727
> .

Commit:
Make all headers constants.

Commit:
Fix comment nit.

Commit:
Change null handling.

harrah commented:
The last comment I have is that a proposed fix for #979 is to version the incremental compiler cache.  Do you have an opinion on how to version it now that we are changing to a text format?  A version header or include it in the file name?

benjyw commented:
I like that idea. Happy to add a version header inside the file. I don't
like relying on file names for such things.

On Thu, Nov 14, 2013 at 5:29 PM, Mark Harrah notifications@github.comwrote:

> The last comment I have is that a proposed fix for #979https://github.com/sbt/sbt/issues/979is to version the incremental compiler cache. Do you have an opinion on how
> to version it now that we are changing to a text format? A version header
> or include it in the file name?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/sbt/sbt/pull/985#issuecomment-28540497
> .

harrah commented:
Ok, thanks.  That + Stamps.toString/fromString and we are set.

Commit:
Minor implementation change.

Per code review comment.

benjyw commented:
What behavior to you want when the version in the file isn't compatible
with the one the format code can handle? Throw a ReadException?

On Thu, Nov 14, 2013 at 5:34 PM, Mark Harrah notifications@github.comwrote:

> Ok, thanks. That + Stamps.toString/fromString and we are set.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/sbt/sbt/pull/985#issuecomment-28540712
> .

harrah commented:
Yes, I think that should work.  There used to be code that would trap any exception while reading from the cache and return an empty Analysis.  As long as that code is still triggered, a ReadException is good.

Commit:
1) Add versioning to analysis file.

2) Refactor Stamp to/from string methods.

benjyw commented:
Ya, that code is still there in FileBasedStore.

OK, pushed a commit with both those changes. I think I've addressed
everything?

On Thu, Nov 14, 2013 at 6:10 PM, Mark Harrah notifications@github.comwrote:

> Yes, I think that should work. There used to be code that would trap any
> exception while reading from the cache and return an empty Analysis. As
> long as that code is still triggered, a ReadException is good.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/sbt/sbt/pull/985#issuecomment-28542115
> .

harrah commented:
Yep, LGTM.  Thanks!

gkossakowski commented:
I looked at all new commits and LGTM as well.

I was wondering, since we have `TestCaseGenerators` that define scalacheck generators for Analysis object would it make sense to write a simple scalacheck test that checks whether for `a: Analysis` you have `read(write(a)) == a`? This could come as a separate PR. @benjyw, wdyt?

For now, let's squash all commits into one and I think we are ready to merge.

benjyw commented:
Closed in favor of #988.

benjyw commented:
Sent a new pull request with all the squashes. And yes - unit test for
load/save would be a good idea.

On Fri, Nov 15, 2013 at 10:42 AM, Grzegorz Kossakowski <
notifications@github.com> wrote:

> I looked at all new commits and LGTM as well.
> 
> I was wondering, since we have TestCaseGenerators that define scalacheck
> generators for Analysis object would it make sense to write a simple
> scalacheck test that checks whether for a: Analysis you have read(write(a))
> == a? This could come as a separate PR. @benjywhttps://github.com/benjyw,
> wdyt?
> 
> For now, let's squash all commits into one and I think we are ready to
> merge.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/sbt/sbt/pull/985#issuecomment-28592927
> .
